src/advanced_inv.cpp

```diff
@@ -2040,23 +2040,37 @@ bool advanced_inventory::query_charges( aim_location destarea, const advanced_in
         redraw = true;
         return false;
     }
+    
     // Check volume, this should work the same for inventory, map and vehicles, but not for worn
     if( unitvolume > 0 && ( unitvolume * amount ) > free_volume && squares[destarea].id != AIM_WORN ) {
-        const long volmax = free_volume / unitvolume;
-        if( volmax <= 0 ) {
+        long room_for;
+        if( it.volume() == 0 ) {
+            // No item should have 0 volume, but I think some of them still do.
+            room_for = item::INFINITE_CHARGES;
+        } else if( !by_charges ) {
+            room_for = free_volume / it.volume();
+        } else {
+            room_for = it.charges_per_volume( free_volume );
```

Comment:
`charges_per_volume` handles item with 0 volume (returns `INFINITE_CHARGES`) and items not counted by charges (same logic as here). You can replace the whole if-else block with a single call to it.